cryptoki/src/mechanism/mod.rs

```diff
@@ -298,6 +300,30 @@ impl MechanismType {
     /// HKDF-DATA mechanism
     pub const HKDF_DATA: MechanismType = MechanismType { val: CKM_HKDF_DATA };
 
+    /// Create vendor defined mechanism
+    ///
+    /// # Arguments
+    ///
+    /// * `adding` - The adding based on `CKM_VENDOR_DEFINED`
+    ///
+    /// Usually vendors defines custom mechanism like this:
+    /// ```c
+    /// #define CKM_SOME_CUSTOM_MECH (CKM_VENDOR_DEFINED | 0x00000001UL)
+    /// ```
+    ///
+    /// It maps to
+    /// ```rust
+    /// use cryptoki::mechanism::MechanismType;
+    ///
+    /// pub const CKM_SOME_CUSTOM_MECH: MechanismType =
+    ///     MechanismType::new_vendor_defined(0x00000001);
+    /// ```
+    pub const fn new_vendor_defined(adding: CK_ULONG) -> MechanismType {
```

Comment:
Should we shortcut it and directly allow passing the mechanism type value (like directly `CKM_SOME_CUSTOM_MECH` in your example). The function could though have a small safety check and check that the mechanism type is higher or equal than `CKM_VENDOR_DEFINED`.

Also, what do you think of that function being `unsafe`? Since we are short circuiting the abstraction layers built in this crate?

Reply:
Hi, thanks for reviewing.

The reason for using `adding` here is to avoid returning a `Result` or `Option` that can not be unwraped in constants. If we want to change this while keeping it as a const function, we will have to use `panic!`.

Regarding the unsafety, my understanding is there are 2 sides.
1. User side: We are (kind of?) sure the user's input is valid by lifetime.
2. Vendor side: I think the current implementation already assumes vendors are trustworthy (by not marking functions as unsafe). So maybe we should continue trusting their library will return reasonable results without breaking safety rules?

Reply:
> keeping it as a const function

Is it something which is needed to make this implementation work? Maybe a normal function is also ok?

I agree with your two points but was thinking what could happen if a completely wrong `adding` was given to the `new_vendor_defined` function? Then we would have a instance of `MechanismType` which might lead to errors without having need to have an `unsafe` call. But maybe as long as it's errors and not panics/seg faults it's fine?

Reply:
Const function is not necessary but definitely more user friendly. But I'm ok with directly passing the value, it also makes sense to me.

Yes, I think it's fine, it should return `CKR_MECHANISM_INVALID` and `CKR_MECHANISM_PARAM_INVALID`.

Reply:
> Yes, I think it's fine, it should return CKR_MECHANISM_INVALID and CKR_MECHANISM_PARAM_INVALID.

Nice, let's go with that then!

Reply:
Sure. One last thing is do we consider re-export `CKM_VENDOR_DEFINED` from `cryptoki-sys`?